### PR TITLE
Support external types as custom types

### PR DESCRIPTION
**Pull request by kenchris**
@Honry 

Something like this @marcoscaceres

***
<a href="https://pr-preview.s3.amazonaws.com/kenchris/web-nfc/pull/278.html" title="Last updated on Aug 6, 2019, 12:48 PM UTC (2ee3a2b)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/web-nfc/278/64a26d5...kenchris:2ee3a2b.html" title="Last updated on Aug 6, 2019, 12:48 PM UTC (2ee3a2b)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -1366,34 +1366,44 @@ <h2>The <dfn>NDEFRecordType</dfn> enum</h2>
       </p>
```
Comment (Honry) [inline index.html:1366]: `<h2>The <dfn>NDEFRecordType</dfn> enum</h2>`

It's not an enum anymore.

File: index.html [hunk 2]
```diff
@@ -1366,34 +1366,44 @@ <h2>The <dfn>NDEFRecordType</dfn> enum</h2>
...
         <dd>
-          The enum value representing <a>NDEFRecord</a> of a <a href="#dfn-json">json</a> type.
+          The value representing <a>NDEFRecord</a> of a <a href="#dfn-json">json</a> type.
         </dd>
         <dt><dfn>opaque</dfn></dt>
         <dd>
-          The enum value representing <a>NDEFRecord</a> of an <a href="#dfn-opaque">opaque</a> type.
+          The value representing <a>NDEFRecord</a> of an <a href="#dfn-opaque">opaque</a> type.
+        </dd>
+        <dt><dfn>External type</dfn></dt>
```
Comment (Honry) [inline index.html]: We may need add the external type to the [data mapping](https://w3c.github.io/web-nfc/#data-mapping) table.

File: index.html [hunk 3]
```diff
@@ -1366,34 +1366,44 @@ <h2>The <dfn>NDEFRecordType</dfn> enum</h2>
...
+        <dd>
+          A {{DOMString}} representing a custom type for the <a>external type record</a>.
+          The type must follow the [[NFC-RTD]] and thus, be a URN with the NID of "`nfc`"
```
Comment (Honry) [inline index.html]:
```suggestion
          The type must follow the [[[NFC-RTD]]] and thus, be a URN with the NID of "`nfc`"
```

File: index.html [hunk 4]
```diff
@@ -3164,19 +3164,19 @@ <h3>Receiving and parsing content</h3>
               Using the <a data-cite="encoding#encoder">encoder</a>, it is only possible
               to encode as UTF-8, unless you do the encoding manually
               and use the "`opaque`"
-              <a href="#the-ndefrecordtype-enum">recordType</a>, which
+              <a href="#the-ndefrecordtype-type">recordType</a>, which
```
Comment (marcoscaceres) [inline index.html]: note, you should never need `href=`... ReSpec should do that.

**Comment by marcoscaceres** [comment]
This seems self contradicting... an enum is a white list, but the you have a DOMString that basically says anything goes.

File: index.html (hunk 2 above)
Comment (kenchris) [reply index.html]: That is done, try the preview

**Comment by marcoscaceres** [comment]
I think you basically just want to take a DOMString, then have a case statement that checks the string values... and throws a "NotSupportedError" when the passed value is unknown.

**Comment by kenchris** [comment]
OK, so I just make it a string... but I still want to list common standardized values

**Comment by marcoscaceres** [comment]
another question, can this be used as fingerprinting entropy? That is, can I poke at the implementation to find out what urns recognizes? And is there are a potential that the urns will vary from one device to another?

**Comment by kenchris** [comment]
No, all implementations should support the same URNs...

**Comment by marcoscaceres** [comment]
ok, cool... then why not just all the urns in the enum?

File: index.html [hunk 5]
```diff
@@ -1358,42 +1358,56 @@ <h4>
...
-          "opaque"
-        };
+        typedef DOMString NDEFRecordType;
       </pre>
+      <p>
+        A set of known standardized values exists, but it is also possible
+        for organizations to create their own custom <dfn>external type</dfn>s,
+        following the following ABNF (specialization of URN) as explained in
+        the [[[NFC-RTD]]]:
+        <pre>
```
Comment (marcoscaceres) [inline index.html]:
```suggestion
        <pre class="abnf">
```

**Comment by kenchris** [comment]
So talked with riju on chat and he approved this change

**Comment by reillyeon** [comment]
> ok, cool... then why not just all the urns in the enum?

@kenchris did you have an answer for this?

**Comment by kenchris** [comment]
There are an infinite amount as they are user definable

**Comment by reillyeon** [comment]
> There are an infinite amount as they are user definable

Thanks for clarifying. Some values of `recordType` are handled specially but all types can be supported by all implementations.

**Pull request merged**